travis: switch from Ubuntu 14.04 to 16.04

mhinz opened the pull request:
Travis is phasing out its support for containers, so we remove `sudo:
false`, which will be a no-op soon.

Reference:

  https://blog.travis-ci.com/2018-11-19-required-linux-infrastructure-migration

mhinz committed:
travis: switch from Ubuntu 14.04 to 16.04

Travis is phasing out its support for containers, so we remove the `sudo:
false`, which will be a no-op soon.

Reference: https://blog.travis-ci.com/2018-11-19-required-linux-infrastructure-migration

Changes for Linux:

  - Xenial comes with libtool installed already. It only provides "libtoolize",
    though. For "libtool" we need to install libtool-bin.

mhinz committed:
xenial: fix clang error messages

* Remove FUNC_ATTR_NONNULL_ALL from function without pointer arguments.

  Otherwise the ASAN build would complain:

  error: 'nonnull' attribute applied to function with no pointer arguments [-Werror,-Wignored-attributes]
  static void do_autocmd_focusgained(_Bool gained) __attribute__((nonnull));

mhinz commented:
Currently blocked by https://github.com/neovim/neovim/issues/8274

jamessan commented:
bot-ci should be updated when this merges too.  There's not a hard requirement to keep the linux dists in sync between the two repos, like there is with macOS, but it's cleaner to do so.

jamessan commented on CMakeLists.txt:
```diff
@@ -251,7 +251,7 @@ if(MSVC)
 else()
   add_definitions(-Wall -Wextra -pedantic -Wno-unused-parameter
-    -Wstrict-prototypes -std=gnu99)
+    -Wno-tautological-compare -Wstrict-prototypes -std=gnu99)
```
It would be better if we could limit this to the files we know might trigger this warning, or fix the macros to not trigger it.

jamessan replied:
This is a false positive in the version of clang-7 that's currently in Xenial.  Since we've switched to clang-4.0, I removed this.  When we switch back to a newer clang, this won't be an issue.

jamessan commented on .travis.yml:
```diff
@@ -50,8 +50,9 @@ jobs:
   include:
     - stage: normal builds
       os: linux
-      compiler: clang
```
I think you can set `clang-4.0` here to achieve the same effect.

mhinz replied:
I tried that, but cmake would always complain that the sanitizer can only be used with clang. I think the `compiler` setting just sets `$CC` which doesn't seem to be enough. But even with `CMAKE_C_COMPILER` set I get that message. ;x

jamessan replied:
That's because you don't have _any_ compiler set, so it [gets set to gcc](https://travis-ci.org/neovim/neovim/jobs/457719820#L798-L799).

mhinz replied:
Oh, I see. My `$CC` got overruled by the default `compiler`.

jamessan committed:
Remove extraneous parens to silence -Wparentheses-equality

    In file included from ../src/nvim/eval/encode.c:974:
    ../src/nvim/eval/typval_encode.c.h:390:40: error: equality comparison with extraneous parentheses [-Werror,-Wparentheses-equality]
              do { if ((tv->vval.v_special == kSpecialVarTrue)) { msgpack_pack_true(packer); } else { msgpack_pack_false(packer); } } while (0);
                        ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~

jamessan committed:
Downgrade to clang-4.0 to avoid false-positive warnings from clang

* -Wtautological-compare

    error: self-comparison always evaluates to true [-Werror,-Wtautological-compare]
    for (win_T *wp = ((curtab) == curtab) ? firstwin : (curtab)->tp_firstwin; wp != ((void*)0); wp = wp->w_next) {

* -Wconversion

    error: implicit conversion loses floating-point precision: 'const float_T' (aka 'const double') to 'float' [-Werror,-Wconversion]
          do { const float_T flt_ = (tv->vval.v_float); switch ((sizeof (flt_) == sizeof (float) ? __fpclassifyf (flt_) : sizeof (flt_) == sizeof (double) ? __fpclassify (flt_) : __fpclassifyl (flt_))) { case 0: { ga_concat(gap, (char_u *)(char_u *) "str2float('nan')"); break; } case 1: { if (flt_ < 0) { ga_append(gap, '-'); } ga_concat(gap, (char_u *)(char_u *) "str2float('inf')"); break; } default: { char numbuf[NUMBUFLEN]; vim_snprintf(numbuf, ((sizeof(numbuf)/sizeof((numbuf)[0])) / ((size_t)(!(sizeof(numbuf) % sizeof((numbuf)[0]))))), "%g", flt_); ga_concat(gap, (char_u *)(char_u *) numbuf); } } } while (0);
                                                                                                   ~~~~~~~~~~~~~  ^~~~

justinmk commented on ci/install.sh:
```diff
@@ -8,10 +8,9 @@ if [[ "${CI_TARGET}" == lint ]]; then
 
 if [[ "${TRAVIS_OS_NAME}" == osx ]]; then
-  brew install ninja
-  brew install gettext
```
so we don't need gettext anymore?

mhinz replied:
The image is provisioned with the latest version of gettext already.

From the latest master build (which still uses the same macOS image): https://travis-ci.org/neovim/neovim/jobs/457894739#L150

justinmk replied:
do you plan to remove `brew install libtool` ?

mhinz replied:
Done!

blueyed commented:
Just for reference: downgrading clang only worked around a ccache issue (which can be fixed through config, and is on by default in newer versions (CCACHE_CPP2)).  https://github.com/neovim/neovim/pull/10487
